src/common-util/src/main/java/ee/ria/xroad/common/util/IsSubjectClientIdDecoder.java

```diff
@@ -0,0 +1,156 @@
+package ee.ria.xroad.common.util;
+
+import ee.ria.xroad.common.CodedException;
+import ee.ria.xroad.common.ErrorCodes;
+import ee.ria.xroad.common.identifier.ClientId;
+
+import org.bouncycastle.asn1.x500.X500Name;
+import org.bouncycastle.asn1.x500.style.BCStyle;
+
+import javax.security.auth.x500.X500Principal;
+
+import java.security.cert.X509Certificate;
+import java.util.regex.Pattern;
+
+import static ee.ria.xroad.common.util.CertUtils.getRDNValue;
+
+/**
+ * Helper class for decoding ClientId from Finnish X-Road instance signing
+ * certificates. adapted to Iceland 060718 JJB It ráðgjöf ehf.
+ */
+public final class IsSubjectClientIdDecoder {
+
+    public static final int NUM_COMPONENTS = 3;
+
+    private IsSubjectClientIdDecoder() {
+        // utility class
+    }
+
+    /**
+     * @param cert certificate from which to construct the client ID
+     * @return a fully constructed Client identifier from DN of the certificate.
+     */
+    public static ClientId getSubjectClientId(X509Certificate cert) {
+        X500Principal principal = cert.getSubjectX500Principal();
+        X500Name x500name = new X500Name(principal.getName());
+
+        if (getRDNValue(x500name, BCStyle.SERIALNUMBER) == null) {
+            if (getRDNValue(x500name, BCStyle.OU) == null) {
+                return CertUtils.getSubjectClientId(cert);
+            }
+            return parseClientIdFromLegacyName(x500name);
+        }
+        return parseClientId(x500name);
+    }
+
+    /*
+     * The encoding for clientID: <ul> <li>C = FI (country code must be 'FI' when
+     * using this decoder)</li> <li>O = organization (must be present) <li>CN =
+     * memberCode (business code without "Y" prefix)</li> <li>serialNumber =
+     * instanceIdentifier;serverCode;memberClass </ul>
+     */
+
+    private static final Pattern SPLIT_PATTERN = Pattern.compile("=");
+
+    private static ClientId parseClientId(X500Name x500name) {
+        String c = getRDNValue(x500name, BCStyle.C);
+        if (!"IS".equals(c)) {
+            throw new CodedException(ErrorCodes.X_INCORRECT_CERTIFICATE,
+                    "Certificate subject name does not contain valid country code");
+        }
+
+        if (getRDNValue(x500name, BCStyle.O) == null) {
+            throw new CodedException(ErrorCodes.X_INCORRECT_CERTIFICATE,
+                    "Certificate subject name does not contain organization");
+        }
+
+        String memberCode = getRDNValue(x500name, BCStyle.CN);
+        if (memberCode == null) {
+            throw new CodedException(ErrorCodes.X_INCORRECT_CERTIFICATE,
+                    "Certificate subject name does not contain common name");
+        }
+        /* added 060718 JJB */
+        String ou = getRDNValue(x500name, BCStyle.OU);
+        if (ou == null) {
+            throw new CodedException(ErrorCodes.X_INCORRECT_CERTIFICATE, "Certificate OU incorrect");
+        }
+
+        String serialNumber = getRDNValue(x500name, BCStyle.SERIALNUMBER);
```

Comment:
It seems that serialNumber is not used, so it could be omitted from the profile.

Reply:
I will look into this on monday morning.

Reply:
This is identical implimentation as in src/common-util/src/main/java/ee/ria/xroad/common/util/FISubjectClientIdDecoder.java . I suggest we we include the ```serualNumber``` for now.